Source/Parallelization/WarpXComm_K.H

```diff
@@ -641,118 +486,181 @@ void warpx_interp_nd_efield_z (int j, int k, int l,
-#if (AMREX_SPACEDIM == 2)
-    Eza(j,k,0) = 0.5_rt*(Ezf(j,k-1,0) + Ezf(j,k,0));
-    amrex::ignore_unused(l);
-#else
-    Eza(j,k,l) = 0.5_rt*(Ezf(j,k,l-1) + Ezf(j,k,l));
+    // Staggering (s = 0 if cell-centered, s = 1 if nodal)
+    const int sj = arr_stag[0];
+    const int sk = arr_stag[1];
+#if (AMREX_SPACEDIM == 3)
+    const int sl = arr_stag[2];
 #endif
-}
 
-// With the PSATD Maxwell solver, this is the arbitrary-order interpolation function used to
-// interpolate the Ez field from a staggered grid to a nodal grid, before gathering
-// the field from the nodes to the particle positions (momentum-conserving gathering)
-AMREX_GPU_DEVICE AMREX_FORCE_INLINE
-void warpx_interp_nd_efield_z (int j, int k, int l,
-                               amrex::Array4<amrex::Real> const& Eza,
-                               amrex::Array4<amrex::Real const> const& Ezf,
-                               const int noz,
-                               amrex::Real const* stencil_coef_z)
-{
-    using namespace amrex;
-#if (AMREX_SPACEDIM == 2)
-    amrex::Real res = 0._rt;
-    for (int nz = 1; nz < noz/2+1; nz++) {
-        res += 0.5_rt * stencil_coef_z[nz-1] * (Ezf(j, k + nz - 1, l) + Ezf(j, k - nz, l));
+    // Number of points (+1) used for interpolation from the staggered grid to the nodal grid
+    // (if the input field is nodal along a given direction, then s = 1, the variable n = 0
+    // is actually never used, and there is no interpolation, only a copy of one single value)
+    const int nj = (sj == 0) ? nox/2 + 1 : 0;
+#if   (AMREX_SPACEDIM == 2)
+    const int nk = (sk == 0) ? noz/2 + 1 : 0;
+#elif (AMREX_SPACEDIM == 3)
+    const int nk = (sk == 0) ? noy/2 + 1 : 0;
+    const int nl = (sl == 0) ? noz/2 + 1 : 0;
+#endif
+
+    // Additional normalization factor
+    const amrex::Real wj = (sj == 0) ? 0.5_rt : 1.0_rt;
+    const amrex::Real wk = (sk == 0) ? 0.5_rt : 1.0_rt;
+#if   (AMREX_SPACEDIM == 2)
+    constexpr amrex::Real wl = 1.0_rt;
+#elif (AMREX_SPACEDIM == 3)
+    const amrex::Real wl = (sl == 0) ? 0.5_rt : 1.0_rt;
+#endif
+
+    // Auxiliary variables for stencil coefficients
+#ifdef WARPX_USE_PSATD
+    amrex::Real cj = 1.0_rt;
+    amrex::Real ck = 1.0_rt;
+    amrex::Real cl = 1.0_rt;
+    int dj, dk;
+#if (AMREX_SPACEDIM == 3)
+    int dl;
+#endif
+#endif
+
+    // Min and max for interpolation loop along j
+    const int jmin = (sj == 0) ? j - nj + 1 : j;
+    const int jmax = (sj == 0) ? j + nj - 2 : j;
+
+    // Min and max for interpolation loop along k
+    const int kmin = (sk == 0) ? k - nk + 1 : k;
+    const int kmax = (sk == 0) ? k + nk - 2 : k;
+
+    // Min and max for interpolation loop along l
+#if   (AMREX_SPACEDIM == 2)
+    // l = 0 always
+    const int lmin = l;
+    const int lmax = l;
+#elif (AMREX_SPACEDIM == 3)
+    const int lmin = (sl == 0) ? l - nl + 1 : l;
+    const int lmax = (sl == 0) ? l + nl - 2 : l;
+#endif
+
+    amrex::Real res = 0.0_rt;
+
+#ifndef WARPX_USE_PSATD // FDTD (linear interpolation)
```

Comment:
I decided to implement two alternative code blocks for the interpolation loops with FDTD and PSATD, in order to have a cleaner and more readable code for the more complex PSATD case below.